File: LeetCode/DataStructure_II/Day18/236_lowest_common_ancestor_of_a_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def DFS(self, node, p, q, paths, path):
        if node is None: return 
        path.append(node)
        if node.val == p.val or node.val == q.val:
            paths.append(path[:])
        self.DFS(node.left, p, q, paths, path)
        self.DFS(node.right, p, q, paths, path)
        path.pop()
        
            
        
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        paths = []
        self.DFS(root, p, q, paths, [])
        idx = -1
        for elem1, elem2 in zip(paths[0], paths[1]):
            if not elem1.val == elem2.val:
                break
            idx +=1 
        return paths[0][idx]
        

File: LeetCode/DayChallenges/DayChallenge_2022_July/236_Lowest_Common_Ancestor_of_a_Binary_Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */
 var lowestCommonAncestor = function(root, p, q) {
    let result;
    check(root);
    return result;
    function check (node) {
        if (node === null) return false;
        const leftExist = check(node.left);
        const rightExist = check(node.right);
        const exist = q === node || p === node;
        if (leftExist || rightExist) {
            if ((leftExist && exist) || (leftExist && rightExist) || (rightExist && exist)) {
                if (!result) {
                    result = node;    
                }
            }
            return true;
        }
        if (exist) {
            return true;
        }
        return false;
    };
};

